crates/core/tedge_mapper/src/c8y/dynamic_discovery.rs

```diff
@@ -0,0 +1,128 @@
+use std::{ffi::OsString, path::PathBuf};
+
+use inotify::{Event, EventMask, Inotify, WatchMask};
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize, Debug)]
+pub enum EventType {
+    ADD,
+    REMOVE,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct DiscoverOp {
+    pub ops_dir: PathBuf,
+    pub event_type: EventType,
+    pub operation_name: String,
+}
+
+#[derive(thiserror::Error, Debug)]
+#[allow(clippy::enum_variant_names)]
+pub enum DynamicDiscoverOpsError {
+    #[error("No Operations directory found: {0}")]
+    NoOperations(String),
+
+    #[error("No event")]
+    NoEventName,
+
+    #[error("Inotify event {0} not supported")]
+    UnsupportedEvent(String),
+
+    #[error(transparent)]
+    EventError(#[from] std::io::Error),
+}
+
+pub fn create_inotify_watch(ops_dir: Option<PathBuf>) -> Result<Inotify, DynamicDiscoverOpsError> {
+    match ops_dir {
+        Some(dir) => {
+            let mut inotify = Inotify::init()?;
+            inotify
+                .add_watch(dir.clone(), WatchMask::CLOSE_WRITE | WatchMask::DELETE)
+                .map_err(|_| {
+                    DynamicDiscoverOpsError::NoOperations(dir.to_string_lossy().to_string())
```

Comment:
The error case name is really confusing. One might have operations, but for some reason one fails to watch them. Simply return the `io::Error`.